NOTES-daneroo.md

```diff
@@ -0,0 +1,104 @@
+
+The timeouts were mostly coming from the `navLeft()` iterator, so I disabled actual downloading and focused on traversal. (`-list` option). I also added an option (`-all`) to bypass the incremental behavior of `$dldir/.lastDone`. I then added `-vt (VerboseTiming`, to isolate timing metrics reporting specifically.
+
+The last main issue was the degradation of performance over time. Probably due to resource leaks in the webapp, especially when iterating at high speed. I found that the simplest solution to this was simply to force a reload every 1000 iterations. This typically takes about `1s` and avoids the `10X` slowdown I was observing.
+
+The following are details of each of the major changes.
+
+## `navLeft()`
+
+- By adding the `-list` option, its functionality is isolated.
+- Remove `WaitReady("body",..)`, as it has no effect (The document stays loaded when we navigate from image to image), and replace it with a loop (until `location != prevLocation`).
+- When only listing, this is the critical part of the loop, so I spent some time experimenting with the timing of the chromedp interactions to optimize total throughput. What seemed to work best was an exponential backoff, so that we can benefit from the *most likely* fast navigation, but don't overtax the browser when we experience the occasional longer navigation delay (which can go up to multiple seconds).
+
+- We also pass in `prevLoaction` from `navN()`, to support the new `navLeft()` loop, to optimize the throughput.
+
+## `navN()`
+
+- Addition of timing code
+- Support listing only (`-list`)
+- Reload page every (`batchSize=1000`) items.
+
+## Termination Criteria
+
+We explicitly fetch the lastPhoto, on the main album page (immediately after authentication), and rely on that, instead of `navLeft()` failing to navigate (`location == prevLocation`). The lastPhoto is determined with a CSS selector evaluation (`a[href^="./photo/"]`), which is fast and should be stable over time, as it uses the `<a href=.. >` detail navigation mechanism of the main album page. This opens another possibility for iteration on the album page itself. (see `listFromAlbum` below)
+
+Another issue, although the `.lasDone` being captured on a successful run is useful, as photos are listed in EXIF date order. If older photos are added to the album they would not appear in subsequent runs. So it would be useful in general to rerun over the entire album (`-all`).
+
+## Headless
+
+If we don't need the authentication flow, (and we persist the auth creds in the profile `-dev`), Chrome can be brought up in `-headless` mode, which considerably reduces memory footprint (`>2Gb` to `<500Mb`), and marginally (23%) increases throughput for the listing experiment.
```

Comment:
yes, we definitely want headless mode.
Maybe we just run by default in headless mode, unless the -dev flag is set?
WDYT? you want to do a PR for that?

Reply:
Sorry I've been away for a few days myself for the holidays.

I will start with this one in a separate PR.
- We **can't** do authentication in `-headless` mode, so it only makes sense for -dev
- It can either be the default for -dev, and fail quickly if not authenticated (not wait for 2 minutes)
- I will comment more on options in the PR...

Reply:
ACK